chain/network/src/peer_manager/connection/mod.rs

```diff
@@ -270,7 +282,45 @@ impl Pool {
                     }
                 }
             }
-            pool.ready.insert(id.clone(), peer);
+            if pool.ready.insert(id.clone(), peer.clone()).is_some() {
+                return Err(PoolError::AlreadyConnected);
+            }
+            if let Some(owned_account) = &peer.owned_account {
+                // Only 1 connection per account key is allowed.
+                // Having 2 peers use the same account key is an invalid setup,
+                // which violates the BFT consensus anyway.
+                // TODO(gprusak): an incorrectly closed TCP connection may remain in ESTABLISHED
+                // state up to minutes/hours afterwards. This may cause problems in
+                // case a validator is restarting a node after crash and the new node has the same
+                // peer_id/account_key/IP:port as the old node. What is the desired behavior is
+                // such a case? Linux TCP implementation supports:
+                // TCP_USER_TIMEOUT - timeout for ACKing the sent data
+                // TCP_KEEPIDLE - idle connection time after which a KEEPALIVE is sent
+                // TCP_KEEPINTVL - interval between subsequent KEEPALIVE probes
+                // TCP_KEEPCNT - number of KEEPALIVE probes before closing the connection.
+                // If it ever becomes a problem, we can eiter:
+                // 1. replace TCP with sth else, like QUIC.
+                // 2. use some lower level API than tokio::net to be able to set the linux flags.
+                // 3. implement KEEPALIVE equivalent manually on top of TCP to resolve conflicts.
+                // 4. allow overriding old connections by new connections, but that will require
+                //    a deeper thought to make sure that the connections will be eventually stable
+                //    and that incorrect setups will be detectable.
+                if let Some(conn) = pool.ready_by_account_key.insert(owned_account.account_key.clone(), peer.clone()) {
+                    // Unwrap is safe, because pool.ready_by_account_key is an index on connections
+                    // with owned_account present.
+                    let err = Err(PoolError::AlreadyConnectedAccount{
+                        peer_id: conn.peer_info.id.clone(),
+                        account_key: conn.owned_account.as_ref().unwrap().account_key.clone(),
+                    });
+                    // We duplicate the error logging here, because returning an error
+                    // from insert_ready is expected (pool may regularly reject connections),
+                    // however conflicting connections with the same account key indicate an
+                    // incorrect validator setup, so we log it here as a warn!, rather than just
+                    // info!.
+                    tracing::warn!(target:"network", "Pool::register({id}): {err:?}");
```

Comment:
can you also add a metric to track this?

Reply:
done